src/sentry/sentry_apps/logic.py

```diff
@@ -0,0 +1,463 @@
+
+    def _update_overview(self) -> None:
+        if self.overview is not None:
+            self.sentry_app.overview = self.overview
+
+    def _update_allowed_origins(self) -> None:
+        if self.allowed_origins and self.sentry_app.application:
+            self.sentry_app.application.allowed_origins = "\n".join(self.allowed_origins)
+            self.sentry_app.application.save()
+
+    def _update_popularity(self, user: User) -> None:
+        if self.popularity is not None:
+            if _is_elevated_user(user):
+                self.sentry_app.popularity = self.popularity
+
+    def _update_schema(self) -> set[str] | None:
+        if self.schema is not None:
+            self.sentry_app.schema = self.schema
+            new_schema_elements = self._get_new_schema_elements()
+            self._delete_old_ui_components()
+            self._create_ui_components()
+            return new_schema_elements
+        return None
+
+    def _get_new_schema_elements(self) -> set[str]:
+        current = SentryAppComponent.objects.filter(sentry_app=self.sentry_app).values_list(
+            "type", flat=True
+        )
+        return _get_schema_types(self.schema) - set(current)
+
+    def _delete_old_ui_components(self) -> None:
+        SentryAppComponent.objects.filter(sentry_app_id=self.sentry_app.id).delete()
+
+    def _create_ui_components(self) -> None:
+        if self.schema is not None:
+            for element in self.schema.get("elements", []):
+                SentryAppComponent.objects.create(
+                    type=element["type"], sentry_app_id=self.sentry_app.id, schema=element
+                )
+
+    def record_analytics(self, user: User, new_schema_elements: set[str] | None) -> None:
+        analytics.record(
+            "sentry_app.updated",
+            user_id=user.id,
+            organization_id=self.sentry_app.owner_id,
+            sentry_app=self.sentry_app.slug,
+            created_alert_rule_ui_component="alert-rule-action" in (new_schema_elements or set()),
+        )
+
+
+@dataclasses.dataclass
+class SentryAppCreator:
+    name: str
+    author: str
+    organization_id: int
+    is_internal: bool
+    scopes: list[str] = dataclasses.field(default_factory=list)
+    events: list[str] = dataclasses.field(default_factory=list)
+    webhook_url: str | None = None
+    redirect_url: str | None = None
+    is_alertable: bool = False
+    verify_install: bool = True
+    schema: Schema = dataclasses.field(default_factory=dict)
+    overview: str | None = None
+    allowed_origins: list[str] = dataclasses.field(default_factory=list)
+    popularity: int | None = None
+    metadata: dict | None = field(default_factory=dict)
+
+    def __post_init__(self) -> None:
+        if self.is_internal:
+            assert (
+                not self.verify_install
+            ), "Internal apps should not require installation verification"
+
+    def run(
+        self,
+        *,
+        user: User,
```

Comment:
This was changed from User | RpcUser to just User. Though I wasn't 100% on this one. Like SentryApp creation and updating all happens in the CS so can we assume all 'users'  won't be RpcUsers

Reply:
>  Like SentryApp creation and updating all happens in the CS so can we assume all 'users' won't be RpcUsers ?

We can likely use the narrower type. The RpcUser could show up in a future refactoring, but we can deal with that then.